### PR TITLE
Expose loaded plugins by type

### PR DESCRIPTION
**Pull request by litvinenko-kaltura**
- Player interface - we've added the following method:
    @NonNull <PluginType> List<PluginType> getLoadedPluginsOfType(Class<PluginType> pluginClass);
    Proper annotations were added as well.
- Three classes were implemented this method:
    - PlayerDecoratorBase
    - PlayerController
    - PlayerLoader
- PlayerLoader.class - has an implementation of method getLoadedPluginsOfType()

### CODE REVIEW

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.kaltura.playkit;
+
+public interface OnMediaInterceptorListener {
+    void onApplyMediaCompleted();
+}
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: This class has to move to Kaltura Player.

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.kaltura.playkit;
+
+public interface OnMediaInterceptorListener {
+    void onApplyMediaCompleted();
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: Rename to `onComplete` and add `@Nullable Exception` as parameter.

File: playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.kaltura.playkit;
+
+/**
+ * Created by alex_lytvynenko on 09.11.2020.
+ */
+public interface PKMediaEntryInterceptor {
+    void apply(PKMediaEntry mediaEntry, OnMediaInterceptorListener listener);
+}
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java]: This class has to move to Kaltura Player.

File: playkit/src/main/java/com/kaltura/playkit/Player.java [hunk 4]
```diff
@@ -527,11 +540,18 @@
...
      * @param listener - event listener
      */
     void removeListener(@NonNull PKEvent.Listener listener);
 
+    /**
+     * Get loaded plugins of type.
+     *
+     * @param pluginClass - PluginType class.
+     */
+    @NonNull <PluginType> List<PluginType> getLoadedPluginsOfType(Class<PluginType> pluginClass);
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/Player.java]:
```suggestion
    @NonNull <PluginType> List<PluginType> getLoadedPluginsByType(Class<PluginType> pluginClass);
```

File: playkit/src/main/java/com/kaltura/playkit/PlayerDecoratorBase.java [hunk 5]
```diff
@@ -160,6 +162,12 @@ public void removeListener(@NonNull PKEvent.Listener listener) {
         player.removeListeners(listener);
     }
 
+    @NonNull
+    @Override
+    public  <PluginType> List<PluginType> getLoadedPluginsOfType(Class<PluginType> pluginClass) {
+        return player.getLoadedPluginsOfType(pluginClass);
+    }
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/PlayerDecoratorBase.java]:
```suggestion
    public  <PluginType> List<PluginType> getLoadedPluginsByType(Class<PluginType> pluginClass) {
        return player.getLoadedPluginsByType(pluginClass);
    }
```

File: playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java [hunk 6]
```diff
@@ -249,4 +249,15 @@ public void addListener(Object groupId, Enum type, PKEvent.Listener listener) {
...
     }
+
+    @NonNull
+    @Override
+    public <PluginType> List<PluginType> getLoadedPluginsOfType(Class<PluginType> pluginClass) {
+        List<PluginType> filteredPlugins = new ArrayList<>();
+        for (LoadedPlugin loadedPlugin : loadedPlugins.values()) {
+            if (pluginClass.isAssignableFrom(loadedPlugin.plugin.getClass()))
+                filteredPlugins.add((PluginType) loadedPlugin.plugin);
+        }
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java]: Why did you change from my suggested implementation? I'm not saying it's not correct, just want to know if the other one wasn't good.

File: playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java [hunk 7]
```diff
@@ -249,4 +249,15 @@ public void addListener(Object groupId, Enum type, PKEvent.Listener listener) {
     public void removeListeners(@NonNull Object groupId) {
         messageBus.removeListeners(groupId);
     }
+
+    @NonNull
+    @Override
+    public <PluginType> List<PluginType> getLoadedPluginsOfType(Class<PluginType> pluginClass) {
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java]:
```suggestion
    public <PluginType> List<PluginType> getLoadedPluginsByType(Class<PluginType> pluginClass) {
```

File: playkit/src/main/java/com/kaltura/playkit/player/PlayerController.java [hunk 8]
```diff
@@ -524,6 +525,13 @@ public void removeListener(@NonNull PKEvent.Listener listener) {
         Assert.shouldNeverHappen();
     }
 
+    @NonNull
+    @Override
+    public <PluginType> List<PluginType> getLoadedPluginsOfType(Class<PluginType> pluginClass) {
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/PlayerController.java]:
```suggestion
    public <PluginType> List<PluginType> getLoadedPluginsByType(Class<PluginType> pluginClass) {
```

**Review by noamtamim — CHANGES_REQUESTED** [review]

File: playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java (hunk 6 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java]: `isInstance()` checks only a superclass. But in case of `PKMediaEntryInterceptor `, which is interface, we have to use `isAssignableFrom`, because it checks both superclass and superinterface.

File: playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java (hunk 3 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java]: If we move it to KalturaPlayer, Broadpeak plugin will have to add kaltura-player dependency as well. Because Broadpeak plugin implements this interface. Are you sure we want to do this?

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java (hunk 1 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: This interface is used by Plugin/Interceptor to notify that it finishes its work. So Broadpeak plugin will have to add kaltura-player dependency as well. Do we need to do that?

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java (hunk 2 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: We will change the name to `onComplete`. 
This interface is used by Plugin/Interceptor to notify that it finishes its work. Currently, the `MessageBus` is in charge of error handling for each interceptor. Interceptors not impact the rest of interceptors, and each of interceptors is independent and can't stop the chain of operations. 
So in order to declare exception: do we need to remove `MessageBus` usage from plugin and send any exception from smartlib in `onComplete(Exception)`? If so, do we need to collect all exceptions from a list of Interceptors or just ignore them?

Comment (noamtamim) [reply playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: The errors reported by plugins in the `MessageBus` are generic and async - they can occur at any time during normal operation of the plugin (not as interceptor).
The errors reported via onComplete(Exception) are specific to this use case: the interceptor/plugin has failed to get the updated/fixed playback URL.

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java (hunk 1 above)
Comment (noamtamim) [reply playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: As the interceptors are meaningless for PlayKit, declaring the interfaces in PlayKit is confusing.

File: playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java (hunk 3 above)
Comment (noamtamim) [reply playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java]: See above - it just doesn't belong to PlayKit.

File: playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java (hunk 6 above)
Comment (noamtamim) [reply playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java]: Ok, makes sense. Is there a warning when you cast `loadedPlugin.plugin` to `PluginType`?

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java (hunk 1 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: ok, i will move it to KalturaPlayer

File: playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java (hunk 3 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/PKMediaEntryInterceptor.java]: ok, i will move it to KalturaPlayer

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java (hunk 2 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: ok, i will do it.

File: playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java (hunk 6 above)
Comment (litvinenko-kaltura) [reply playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java]: yes, `Unchecked cast`

File: playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java (hunk 2 above)
Comment (noamtamim) [reply playkit/src/main/java/com/kaltura/playkit/OnMediaInterceptorListener.java]: @litvinenko-kaltura  correction - the implementation should be:
- the listener method name is onComplete()
- it **doesn't** take an error object (for the reasons you mentioned)
- Errors should be logged and potentially posted via the messagebus (like you suggested)

File: playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java (hunk 6 above)
Comment (noamtamim) [reply playkit/src/main/java/com/kaltura/playkit/PlayerLoader.java]: So please suppress this warning on the statement, and add text that explains it (like I did).

**Review by giladna — APPROVED** [review]

**Review by giladna — APPROVED** [review]

**Review by noamtamim — APPROVED** [review]
